plugins/nordic-hid/fu-nordic-hid-archive.c

```diff
@@ -11,15 +11,198 @@
 	FuFirmwareClass parent_instance;
 };
 
 G_DEFINE_TYPE(FuNordicHidArchive, fu_nordic_hid_archive, FU_TYPE_FIRMWARE)
 
+static const gchar *
+fu_nordic_hid_archive_parse_file_get_bootloader_name(JsonObject *obj, GError **error)
+{
+	if (json_object_has_member(obj, "version_B0"))
+		return "B0";
+	if (json_object_has_member(obj, "version_MCUBOOT"))
+		return "MCUBOOT";
+	if (json_object_has_member(obj, "version_MCUBOOT+XIP"))
+		return "MCUBOOT+XIP";
+
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "only B0 and MCUboot bootloaders are supported");
+	return NULL;
+}
+
+static FuFirmware *
+fu_nordic_hid_archive_parse_file_image_create(const gchar *bootloader_name, GError **error)
+{
+	if (g_strcmp0(bootloader_name, "B0") == 0)
+		return g_object_new(FU_TYPE_NORDIC_HID_FIRMWARE_B0, NULL);
+	if (g_strcmp0(bootloader_name, "MCUBOOT") == 0 ||
+	    g_strcmp0(bootloader_name, "MCUBOOT+XIP") == 0)
+		return g_object_new(FU_TYPE_NORDIC_HID_FIRMWARE_MCUBOOT, NULL);
+
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "only B0 and MCUboot bootloaders are supported");
+	return NULL;
+}
+
+static gchar *
+fu_nordic_hid_archive_parse_file_get_board_name(JsonObject *obj,
+						gint64 manifest_ver,
+						GError **error)
+{
+	g_auto(GStrv) board_split = NULL;
+	const gchar *board_name_readout = NULL;
+
+	if (!json_object_has_member(obj, "board")) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "manifest invalid as has no target board information");
+		return NULL;
+	}
+
+	board_name_readout = json_object_get_string_member(obj, "board");
+	if (board_name_readout == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "manifest invalid as has no target board information");
+		return NULL;
+	}
+
+	if (manifest_ver == 0) {
+		/* for manifest "format-version" of "0", the board name is represented only
+		 * by part of the string before the "_" symbol
+		 */
+		board_split = g_strsplit(board_name_readout, "_", -1);
+		if (board_split[0] == NULL) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INVALID_FILE,
+					    "manifest invalid as has no board information");
+			return NULL;
+		}
+		return g_strdup(board_split[0]);
+	}
+
+	/* duplicate string for consistent memory management */
+	return g_strdup(board_name_readout);
+}
+
+static gint
+fu_nordic_hid_archive_parse_file_get_flash_area_id_v1(JsonObject *obj,
+						      const gchar *bootloader_name,
+						      guint files_cnt,
+						      GError **error)
+{
+	const gchar *image_idx_str = NULL;
+	const gchar *slot_str = NULL;
+	gint64 image_idx = -1;
+	gint64 slot = -1;
+
+	/* for MCUboot bootloader with swap, if only a single image is available,
+	 * the "image_index" and "slot" properties may be omitted
+	 */
+	if (g_strcmp0(bootloader_name, "MCUBOOT") == 0 && files_cnt == 1) {
+		return 0;
```

Comment:
is -1 an error or is 0 an error? Please just make this return gboolean and return the slot using `gint *slot` as a parameter. Thanks.

Reply:
Ok, I will return result through a pointer instead.